### PR TITLE
Change signatures that help is showing (revive #25672, fix #20064)

### LINKED ISSUE #20064: Some docstrings always show up (but they should not)
There is a docstring for `Base.length`, which shows up for any input type (but it should not). This makes it <del>impossible</del> very hard to automatically build documentations in case one wants to document a custom `length` function.

```julia
julia> type A end

julia> doc"""asdasds""" function Base.length(a::A) end
Base.length

help?> length(A())
  length(collection) -> Integer

  For ordered, indexable collections, the maximum index i for which getindex(collection, i) is valid.
  For unordered collections, the number of elements.

  asdasds
```

Happens also for builtin types:
```julia
help?> length("string")
  length(collection) -> Integer

  For ordered, indexable collections, the maximum index i for which getindex(collection, i) is valid.
  For unordered collections, the number of elements.

  length(s)

  The number of characters in string s.
```

### PR DESCRIPTION
**Pull request by thofma**
This is a rebased version of #25672, which fixes the regression noticed in #20064.

I hope you don't mind @KristofferC 

I used the check suggested by Jameson and added tests.

Here is the old and new behavior

### 1.0/master
```julia
help?> +(::Dates.Date, ::Dates.Time)
  +(x, y...)

  Addition operator. x+y+z+... calls this function with all arguments, i.e. +(x, y, z, ...).

  Examples
  ≡≡≡≡≡≡≡≡≡≡

  julia> 1 + 20 + 4
  25
  
  julia> +(1, 20, 4)
  25

  ───────────────────────────────────────────────────────────────────────────────────────────────────────────

  dt::Date + t::Time -> DateTime

  The addition of a Date with a Time produces a DateTime. The hour, minute, second, and millisecond parts of
  the Time are used along with the year, month, and day of the Date to create the new DateTime. Non-zero
  microseconds or nanoseconds in the Time type will result in an InexactError being thrown.
```

### This PR
```julia
help?> +(::Dates.Date, ::Dates.Time)
  dt::Date + t::Time -> DateTime

  The addition of a Date with a Time produces a DateTime. The hour, minute, second, and millisecond parts of
  the Time are used along with the year, month, and day of the Date to create the new DateTime. Non-zero
  microseconds or nanoseconds in the Time type will result in an InexactError being thrown.
```

Consider
```julia
"""
foobar
"""
function foobar end

"""
Number
"""
foobar(x::Number) = x

"""
Real
"""
foobar(x::Real) = x


"""
Int
"""
foobar(x::Int) = x


"""
Real, Int
"""
foobar(x::Real, y::Int) = x
```

### 1.0/master
```julia
help?> foobar(1)
  Number

  ───────────────────────────────────────────────────────────────────────────────────────────────────────────

  Real

  ───────────────────────────────────────────────────────────────────────────────────────────────────────────

  Int

help?> foobar(::Real)
  Number

  ───────────────────────────────────────────────────────────────────────────────────────────────────────────

  Real

help?> foobar(::Real, ::Real)
  Number

  ───────────────────────────────────────────────────────────────────────────────────────────────────────────

  Real

  ───────────────────────────────────────────────────────────────────────────────────────────────────────────

  Int

  ───────────────────────────────────────────────────────────────────────────────────────────────────────────

  Real, Int
```

### This PR
```julia
help?> foobar(1)
  Int

help?> foobar(2.0)
  Real

help?> foobar(2.0, 1)
  Real, Int
help?> foobar(::Real)
  Real

  ───────────────────────────────────────────────────────────────────────────────────────────────────────────

  Int

help?> foobar(::Real, ::Real)
  Real, Int
```

Should be squashed when merged.

### CODE REVIEW

**Comment by thofma** [comment]
Since this fixes a regression/bug introduced in 0.7, can this be backported to 1.0?

File: stdlib/REPL/src/docview.jl [hunk 1]
```diff
@@ -86,15 +86,19 @@ function doc(binding::Binding, sig::Type = Union{})
...
     # Lookup `binding` and `sig` for matches in all modules of the docsystem.
     for mod in modules
         dict = meta(mod)
         if haskey(dict, binding)
             multidoc = dict[binding]
             push!(groups, multidoc)
             for msig in multidoc.order
-                sig <: msig && push!(results, multidoc.docs[msig])
+                if typeintersect(msig, sig) != Union{}
+                #if (isconcretetype(sig) && sig <: msig) || (!isconcretetype(sig) && msig <: sig) || sig == Union{}
```
Comment (KristofferC) [inline stdlib/REPL/src/docview.jl]: Delete?

**Comment by KristofferC** [comment]
I'm not super comfortable putting this straight into 1.0.1. It has had little testing and I am not sure exactly what effects it will have. Feels better to have this live on master for a while first.

**Comment by KristofferC** [comment]
One thing that would be good is to render the docs with and without this PR and diff them.

**Comment by thofma** [comment]
Thanks for your feedback @KristofferC. You are right about the backporting. It has not been tested enough in the wild.

By diffing them, do you mean posting here the different output as you did in #25672?

**Comment by thofma** [comment]
@KristofferC I added the diff at the top.

**Comment by thofma** [comment]
@KristofferC Thanks. I checked and they are different:
With this PR:
http://www.mathematik.uni-kl.de/~thofmann/build_pr/html/en/base/collections.html#Base.pairs
Without:
http://www.mathematik.uni-kl.de/~thofmann/build/html/en/base/collections.html#Base.pairs

In the documentation source it reads ```@doc Base.pairs```.
I don't know what Documenter is doing, because it works in the REPL properly:
```julia
help?> pairs
search: pairs uppercasefirst Pair partialsort partialsort! partialsortperm partialsortperm! PartialQuickSort

  pairs(collection)

  Return an iterator over key => value pairs for any collection that maps a set of keys to a set of values. This includes arrays, where the keys are the array indices.

  ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

  pairs(IndexLinear(), A)
  pairs(IndexCartesian(), A)
  pairs(IndexStyle(A), A)

  An iterator that accesses each element of the array A, returning i => x, where i is the index for the element and x = A[i]. Identical to pairs(A), except that the style of index can be selected. Also similar to
  enumerate(A), except i will be a valid index for A, while enumerate always counts from 1 regardless of the indices of A.

  Specifying IndexLinear() ensures that i will be an integer; specifying IndexCartesian() ensures that i will be a CartesianIndex; specifying IndexStyle(A) chooses whichever has been defined as the native indexing style for
  array A.

  Mutation of the bounds of the underlying array will invalidate this iterator.

  Examples
  ≡≡≡≡≡≡≡≡≡≡

  julia> A = ["a" "d"; "b" "e"; "c" "f"];
  
  julia> for (index, value) in pairs(IndexStyle(A), A)
             println("$index $value")
         end
  1 a
  2 b
  3 c
  4 d
  5 e
  6 f
  
  julia> S = view(A, 1:2, :);
  
  julia> for (index, value) in pairs(IndexStyle(S), S)
             println("$index $value")
         end
  CartesianIndex(1, 1) a
  CartesianIndex(2, 1) b
  CartesianIndex(1, 2) d
  CartesianIndex(2, 2) e

  See also: IndexStyle, axes.
```

**Comment by mortenpi** [comment]
@thofma I am not actually very familiar the docsystem code in Documenter, but as a general principle, Documenter should behave similarly to the REPL I'd say. If there is / will be a difference between them, please do open an issue in Documenter for this. Or if you'd willing to PR the fix, that would be even more awesome.

Also, if we can offload as much of the docsystem logic to Base, instead of duplicating it in Documenter, that would be good too. There's maybe a reason why Documenter does not use `Docs.doc`, but I think it would be worth looking into if we can get rid of this apparently duplication.

**Comment by thofma** [comment]
I tried to investigate why the built documentation looks different with this PR (it really should not since Documenter does its own thing and is not affected by the changes here). But whenever I tried to debug it using some print statements in Documenter, it gave me the correct result. I think I am hitting some weird bug.

I will look at Documenter.

**Comment by thofma** [comment]
@KristofferC Rebased and the bug is gone. Documenation built is not affected. 

This PR here only affects the REPL helpmode and nothing else, which is explained in the OP and covered by tests.

Is there something else that needs to be addressed?

**Comment by thofma** [comment]
Friendly bump

**Comment by thofma** [comment]
The issue persists with latest master, so again a friendly bump.

**Pull request closed without merging**